src/theme-provider.js

```diff
@@ -1,7 +1,32 @@
+// @flow
 import React, {Component} from 'react'
 import brcast from 'brcast'
+import type {Element} from 'react' // eslint-disable-line no-duplicate-imports
+
+// eslint-disable-next-line no-duplicate-imports
+import type {Broadcast, UnsubscribeFunction} from 'brcast'
+
 import {PropTypes} from './react-compat'
 import {CHANNEL} from './constants'
+
+export type Theme = Object;
+
+export type ThemeProviderContext = {
+  [typeof CHANNEL]: Broadcast<Theme>,
+};
+
+type Props = {
+  theme: Theme,
+  children: string | number | false | null | Array<Element<*>>,
+};
+
+// TODO: {[typeof CHANNEL]: void} is an object with the entry
+// `[CHANNEL]: undefined`, but an empty object should be allowed here, too.
+// Flow complains about the missing property if I type it as {}, though,
+// even though an empty object should still return `undefined` when
+// accessing the missing property.
```

Comment:
This is actually the same issue I run into [here](https://github.com/paypal/glamorous/pull/205/files/6526214b9f057d26db60dd3856b75593c0afa0af#diff-3be2a7cdd58ade04e2acbcef3702d3d7R195).

The issue is that flow doesn't refine a value into a more specific type based on a property value unless both types have the same property.

So this works:
```js
type Success = { error: false, data: any }
type Failure = { error: true, errorMessage: string }

function handleResponse(response: Success | Failure) {
  if (!response.error) {
    // Flow knows that this must be a Success, so it lets you
    // access data.
    return response.data
  }
}
```
but this does not:
```js
type Success = { error: false, data: any }
type Failure = { error: true, errorMessage: string }

function handleResponse(response: Success | Failure) {
  if (!response.errorMessage) {
    // Flow cannot guarantee that response is a success
    return response.data
  }
}
```

This is because object types in flow are not "exact" by default; that is, they're allowed to have extra properties:
```js
type SomethingWithValue = { value: number }

const a: SomethingWithValue = { value: 4 } // this is fine
const b: SomethingWithValue = { value: 5, foo: 6 } // this is fine, too
```

Unfortunately, even if you make your types exact, it looks like flow still can't refine them (`{| ... |}` is "exact object type" syntax):
```js
type Success = {| error: false, data: any |}
type Failure = {| error: true, errorMessage: string |}

function handleResponse(response: Success | Failure) {
  if (!response.errorMessage) {
    // Flow still can't make any guarantees
    return response.data
  }
}
```

That might be a bug, but even if flow could refine exact types based on unrelated property presence, using an exact type in this situation wouldn't be correct, since context can have anything in it.